### ISSUE TITLE
Unable to use both OpenCL and Hexagon in the same pipeline

### ISSUE THREAD

**Issue by bleibig** [halide/Halide#1958]
When generating a pipeline where one function is scheduled on the GPU with .gpu_tile() and Target::OpenCL, and another is scheduled on Hexagon, this error occurs:

```
Internal error at /Users/brian/src/Halide/src/CodeGen_LLVM.cpp:2858 triggered by user code at ./generate.cpp:18:
Condition failed: fn
External function halide_hexagon_initialize_kernels is marked as taking user_context, but is not in the runtime module. Check if runtime_api.cpp needs to be rebuilt.
```

Is this an intended limitation with these targets, or just an issue with a runtime module not being linked in?

**Comment by zvookin** [comment]
Not intentional. Do you have anything actually scheduled on Hexagon in this
code? It's a bug either way, but it might be that there isn't any offloaded
code on Hexagon so certain stuff didn't get put in.

-Z-

**Comment by pranavb-ca** [comment]
What target are you using while building this?

**Comment by pranavb-ca** [comment]
if you have HVX_64 or HVX_128 set as a feature in your target, you should get halide_hexagon_initialize_kernels in your runtime module.
See -> https://github.com/halide/Halide/blob/master/src/LLVM_Runtime_Linker.cpp#L865

So, I am curious to see what your target is.

**Comment by dsharletg** [comment]
That LLVM_Runtime_Linker code looks like the problem: note the control flow is:

        if (t.has_feature(Target::CUDA)) {
        } else if (t.has_feature(Target::OpenCL)) {
        } else if (t.has_feature(Target::OpenGL)) {
        } else if (t.has_feature(Target::OpenGLCompute)) {
        } else if (t.has_feature(Target::Metal)) {
        } else if (t.arch != Target::Hexagon && t.features_any_of({Target::HVX_64, Target::HVX_128})) {
        }

It probably should be:

        if (t.has_feature(Target::CUDA)) {
        } 
        if (t.has_feature(Target::OpenCL)) {
        } 
        if (t.has_feature(Target::OpenGL)) {
        } 
        if (t.has_feature(Target::OpenGLCompute)) {
        } 
        if (t.has_feature(Target::Metal)) {
        } 
        if (t.arch != Target::Hexagon && t.features_any_of({Target::HVX_64, Target::HVX_128})) {
        }

**Comment by bleibig** [comment]
This is the test case I'm using:

```c++
#include "Halide.h"

using namespace Halide;

int main() {
    Func f("f"), g("g");
    Var x("x"), y("y"), xi("xi"), yi("yi");

    f(x, y) = x + y;
    g(x, y) = f(x, y) * 2;

    f.compute_root().hexagon().vectorize(x, 128);
    g.gpu_tile(x, y, xi, yi, 16, 16);

    Target t(Target::Android, Target::ARM, 64);
    t.set_feature(Target::OpenCL);
    t.set_feature(Target::HVX_128);
    g.compile_to_file("generated", g.infer_arguments(), "generated", t);

    return 0;
}
```

So the target is Android arm64. It looks like the hexagon offload module is compiling fine, the error is showing up in the main CPU module. 

@dsharletg I do think that's the issue, all of the GPU targets + hexagon are mutually exclusive with each other. I didn't know if it was designed that way for a reason, but if not, it should be the way you suggest.

**Comment by dsharletg** [comment]
I think your test case is not working because of the issue I pointed at. It's an easy change, we just need to figure out how to enable testing for it. I think we should be able to fix it shortly.

**Comment by zvookin** [comment]
test/correctness/gpu_multi_device.cpp should test this if it is run with a
target that has more than one GPU feature.

-Z-

**Comment by dsharletg** [comment]
Yeah, I don't think the buildbots actually run it with more than one GPU feature though, and I don't think it can work due to the issue in LLVM_Runtime_Linker.cpp treating them as mutually exclusive.

**Comment by abadams** [comment]
See #1961

**Comment by dsharletg** [comment]
Did this get fixed with #1961?

**Comment by bleibig** [comment]
Yes, I'll close this.
